pandas/core/strings.py

```diff
@@ -1737,10 +1743,71 @@ def cons_row(x):
                 return cons(result, name=name, index=index)
 
     @copy(str_cat)
-    def cat(self, others=None, sep=None, na_rep=None):
+    def cat(self, others=None, sep=None, na_rep=None, align=None):
+        from pandas.core.series import Series
+        from pandas.core.frame import DataFrame
+        if align is None and (isinstance(others, Series)
+                              or (isinstance(others, list)
+                                  and any(isinstance(x, Series)
+                                          for x in others))):
+            align = False
+            warnings.warn("A future version of pandas will perform alignment "
+                          "when others is a series (or a list containing a"
+                          "series). To disable alignment (the previous "
+                          "behavior) and silence this warning, pass "
+                          "'align=False'. To enable alignment (the future "
+                          "behavior) and silence this warning, pass "
+                          "'align=True'", FutureWarning, stacklevel=2)
+
         data = self._orig if self._is_categorical else self._data
-        result = str_cat(data, others=others, sep=sep, na_rep=na_rep)
-        return self._wrap_result(result, use_codes=(not self._is_categorical))
+        if align and isinstance(others, Series):
+            # str_cat deals with arrays only
+            data, others = data.align(others, join='outer')
+            data = str_cat(data, others=others, sep=sep, na_rep=na_rep)
+            # cannot use _wrap_result because it refers to self._orig,
+            # which we cannot extend without changing the calling series
+            if self._is_categorical:
+                data = take_1d(data, data.cat.codes)
+            return Series(data, index=others.index)
+        # case where others is a list and all elements are series
+        elif (align and isinstance(others, list)
+                and all(isinstance(x, Series) for x in others)):
+            # first achieve maximum extent of data
+            for x in others:
+                data, _ = data.align(x, join='outer')
+            # then bring elements of others to same size
```

Comment:
woa, why are you adding all of this code????

Reply:
I removed the code additions from ``str_cat`` and ``_get_array_list`` because that was messy and for several other reasons not a good way to do it -- all the changes are now in ``str.cat`` itself.

To be able to align in all the different cases (list of ``Series``, list of ``np.ndarrays``, mixture of both, plus all the other variants), it's necessary to add (some variant of) the code I added. I tried to keep it clean, non-redundant, documented, and nicely recursive. I don't believe the desired functionality can be added with substantially less code (have a look at the test cases). Currently working on the requested doc changes.